### PR TITLE
Feat: IP white list

### PR DESCRIPTION
**Pull request by SunnyBoy-WYH**
image

because no pd，just use java list； when pd ready , we can checkout pd.

tips: 

1. this feat work when auth mode was set.
2. this feat work when white ip status was set to be enable.

test:

update whitelist :
image
check whitelist  :

image

open whitelist :
image

### CODE REVIEW

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpAPI.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+import jakarta.ws.rs.POST;
+import jakarta.ws.rs.PUT;
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.QueryParam;
+import jakarta.ws.rs.core.Context;
+
+@Path("whiteiplist")
+@Singleton
+public class WhiteIpAPI extends API {
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpAPI.java:53]: also keep WhiteIpListAPI?

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpAPI.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+        return ImmutableMap.of("whiteIpList", whiteIpList);
+    }
+
+    @POST
+    @Timed
+    @StatusFilter.Status(StatusFilter.Status.ACCEPTED)
+    @Consumes(APPLICATION_JSON)
+    @Produces(APPLICATION_JSON_WITH_CHARSET)
+    @RolesAllowed("admin")
+    public Map<String, Object> batch(@Context GraphManager manager,
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpAPI.java:74]: rename to updateWhiteIPs?

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpAPI.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+                        "Missing argument: action");
+        E.checkArgument(value instanceof String,
+                        "Invalid action type '%s', must be string",
+                        value.getClass());
+        String action = (String) value;
+        E.checkArgument(StringUtils.isNotEmpty(action),
+                        "Missing argument: action");
+        List<String> existed = new ArrayList<>();
+        List<String> loaded = new ArrayList<>();
+        List<String> illegalIps = new ArrayList<>();
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpAPI.java:94]: `existedIPs, loadedIPs, illegalIPs`

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpAPI.java [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+    public Map<String, Object> batch(@Context GraphManager manager,
+                                     Map<String, Object> actionMap) {
+        E.checkArgument(actionMap != null,
+                        "Missing argument: actionMap");
+        List<String> whiteIpList = manager.authManager().listWhiteIp();
+        Object ips = actionMap.get("ips");
+        E.checkArgument(ips instanceof List,
+                        "Invalid ips type '%s', must be list", ips.getClass());
+        List<String> ipList = (List<String>) ips;
+        Object value = actionMap.get("action");
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpAPI.java:83]: keep a naming style, like `ipListRaw-ipList`, `actionRaw-action`

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpAPI.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+        }
+        manager.authManager().setWhiteIpList(whiteIpList);
+        return result;
+    }
+
+    @PUT
+    @Timed
+    @Produces(APPLICATION_JSON_WITH_CHARSET)
+    @RolesAllowed("admin")
+    public Map<String, Object> update(@Context GraphManager manager,
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpAPI.java:135]: updateStatus

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpAPI.java [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+                                      @QueryParam("status") String status) {
+        LOG.debug("Enable or disable white ip list");
+        E.checkArgument("true".equals(status) ||
+                        "false".equals(status),
+                        "Invalid status, valid status is 'true' or 'false'");
+        boolean open = Boolean.parseBoolean(status);
+        manager.authManager().setWhiteIpStatus(open);
+        Map<String, Object> map = new HashMap<>();
+        map.put("WhiteIpListOpen", open);
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpAPI.java:144]: map.put("status", open);

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpAPI.java [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+        manager.authManager().setWhiteIpList(whiteIpList);
+        return result;
+    }
+
+    @PUT
+    @Timed
+    @Produces(APPLICATION_JSON_WITH_CHARSET)
+    @RolesAllowed("admin")
+    public Map<String, Object> update(@Context GraphManager manager,
+                                      @QueryParam("status") String status) {
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpAPI.java:136]: prefer `boolean enabled`

File: hugegraph-api/src/main/java/org/apache/hugegraph/auth/HugeGraphAuthProxy.java [hunk 8]
```diff
@@ -1568,6 +1568,26 @@ public UserWithRole validateUser(String token) {
             }
         }
 
+        @Override
+        public List<String> listWhiteIp() {
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/auth/HugeGraphAuthProxy.java:1572]: prefer listWhiteIPs

File: hugegraph-api/src/main/java/org/apache/hugegraph/auth/HugeGraphAuthProxy.java [hunk 9]
```diff
@@ -1568,6 +1568,26 @@ public UserWithRole validateUser(String token) {
...
+            return this.authManager.listWhiteIp();
+        }
+
+        @Override
+        public void setWhiteIpList(List<String> whiteIpList) {
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/auth/HugeGraphAuthProxy.java:1577]: prefer setWhiteIPs

File: hugegraph-core/src/main/java/org/apache/hugegraph/auth/AuthManager.java [hunk 10]
```diff
@@ -126,4 +126,12 @@ public interface AuthManager {
...
 
     UserWithRole validateUser(String token);
+
+    List<String> listWhiteIp();
+
+    void setWhiteIpList(List<String> whiteIpList);
+
+    boolean getWhiteIpStatus();
+
+    void setWhiteIpStatus(boolean status);
```
Comment (javeme) [inline hugegraph-core/src/main/java/org/apache/hugegraph/auth/AuthManager.java:136]: prefer enableWhiteIpList

File: hugegraph-core/src/main/java/org/apache/hugegraph/auth/AuthManager.java [hunk 11]
```diff
@@ -126,4 +126,12 @@ public interface AuthManager {
     UserWithRole validateUser(String username, String password);
 
     UserWithRole validateUser(String token);
+
+    List<String> listWhiteIp();
+
+    void setWhiteIpList(List<String> whiteIpList);
+
+    boolean getWhiteIpStatus();
```
Comment (javeme) [inline hugegraph-core/src/main/java/org/apache/hugegraph/auth/AuthManager.java:134]: prefer enabledWhiteIpList

File: hugegraph-core/src/main/java/org/apache/hugegraph/auth/AuthManager.java [hunk 12]
```diff
@@ -126,4 +126,12 @@ public interface AuthManager {
     UserWithRole validateUser(String username, String password);
 
     UserWithRole validateUser(String token);
+
+    List<String> listWhiteIp();
+
+    void setWhiteIpList(List<String> whiteIpList);
```
Comment (javeme) [inline hugegraph-core/src/main/java/org/apache/hugegraph/auth/AuthManager.java:132]: prefer setWhiteIPs

File: hugegraph-core/src/main/java/org/apache/hugegraph/auth/AuthManager.java [hunk 13]
```diff
@@ -126,4 +126,12 @@ public interface AuthManager {
     UserWithRole validateUser(String username, String password);
 
     UserWithRole validateUser(String token);
+
+    List<String> listWhiteIp();
```
Comment (javeme) [inline hugegraph-core/src/main/java/org/apache/hugegraph/auth/AuthManager.java:130]: prefer listWhiteIPs

File: hugegraph-core/src/main/java/org/apache/hugegraph/auth/StandardAuthManager.java [hunk 14]
```diff
@@ -77,6 +76,10 @@ public class StandardAuthManager implements AuthManager {
     private final TokenGenerator tokenGenerator;
     private final long tokenExpire;
 
+    private List<String> ipWhiteList;
```
Comment (javeme) [inline hugegraph-core/src/main/java/org/apache/hugegraph/auth/StandardAuthManager.java:79]: mark final

File: hugegraph-core/src/main/java/org/apache/hugegraph/auth/StandardAuthManager.java [hunk 15]
```diff
@@ -77,6 +76,10 @@ public class StandardAuthManager implements AuthManager {
...
+
+    private Boolean whiteIpStatus;
```
Comment (javeme) [inline hugegraph-core/src/main/java/org/apache/hugegraph/auth/StandardAuthManager.java:81]: prefer boolean ipWhiteListEnabled

File: hugegraph-core/src/main/java/org/apache/hugegraph/auth/StandardAuthManager.java (hunk 14 above)
Comment (SunnyBoy-WYH) [reply hugegraph-core/src/main/java/org/apache/hugegraph/auth/StandardAuthManager.java:79]: mark final maybe we cant change it anymore?

Comment (imbajin) [reply hugegraph-core/src/main/java/org/apache/hugegraph/auth/StandardAuthManager.java:79]: > mark final maybe we cant change it anymore?

for `Object`, `final`/`static` will just fix the **Memory Point Address**, won't influence the **inner value**

Comment (SunnyBoy-WYH) [reply hugegraph-core/src/main/java/org/apache/hugegraph/auth/StandardAuthManager.java:79]: > > > mark final maybe we cant change it anymore?
> > 
> > 
> > for `Object`, `final`/`static` will just fix the **Memory Pointer Address**, won't influence the **inner value**
> > just like the **head pointer** in the linked list
> 
> like this？ we cant change object pointer. so we cant use `this.ipWhiteList = ipWhiteList` in setWhiteIPs func.
> 
> ```
> private final List<String> ipWhiteList;
> 
> public void setWhiteIPs(List<String> ipWhiteList) {
>     this.ipWhiteList.clear();
>     this.ipWhiteList.addAll(ipWhiteList);
> }
> ```

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpListAPI.java [hunk 16]
```diff
@@ -71,75 +71,75 @@ public Map<String, Object> list(@Context GraphManager manager) {
...
+        manager.authManager().setWhiteIPs(whiteIpList);
         return result;
     }
 
     @PUT
     @Timed
     @Produces(APPLICATION_JSON_WITH_CHARSET)
     @RolesAllowed("admin")
-    public Map<String, Object> update(@Context GraphManager manager,
+    public Map<String, Object> updateStatus(@Context GraphManager manager,
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpListAPI.java:135]: also update alignment of line 136

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpListAPI.java [hunk 17]
```diff
@@ -71,75 +71,75 @@ public Map<String, Object> list(@Context GraphManager manager) {
...
                 }
-                whiteIpList.addAll(loaded);
+                whiteIpList.addAll(loadedIPs);
                 break;
             case "remove":
                 LOG.debug("Remove from white ip list");
-                result.put("removed", existed);
-                result.put("nonexistent", loaded);
-                whiteIpList.removeAll(existed);
+                result.put("removed", existedIPs);
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpListAPI.java:117]: also keep `removed_ips` style?

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpListAPI.java [hunk 18]
```diff
@@ -71,75 +71,75 @@ public Map<String, Object> list(@Context GraphManager manager) {
...
+                result.put("nonexistent", loadedIPs);
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpListAPI.java:118]: ditto

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpListAPI.java [hunk 19]
```diff
@@ -71,75 +71,75 @@ public Map<String, Object> list(@Context GraphManager manager) {
     @Consumes(APPLICATION_JSON)
     @Produces(APPLICATION_JSON_WITH_CHARSET)
     @RolesAllowed("admin")
-    public Map<String, Object> batch(@Context GraphManager manager,
+    public Map<String, Object> updateWhiteIPs(@Context GraphManager manager,
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpListAPI.java:74]: also update alignment of line 75

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpListAPI.java [hunk 20]
```diff
@@ -71,75 +71,75 @@ public Map<String, Object> list(@Context GraphManager manager) {
...
+                whiteIpList.addAll(loadedIPs);
                 break;
             case "remove":
                 LOG.debug("Remove from white ip list");
-                result.put("removed", existed);
-                result.put("nonexistent", loaded);
-                whiteIpList.removeAll(existed);
+                result.put("removed", existedIPs);
+                result.put("nonexistent", loadedIPs);
+                whiteIpList.removeAll(existedIPs);
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpListAPI.java:119]: try to set the type of whiteIpList to ` Set<String>`

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpListAPI.java [hunk 21]
```diff
@@ -71,75 +71,75 @@ public Map<String, Object> list(@Context GraphManager manager) {
...
         }
         switch (action) {
             case "load":
                 LOG.debug("Load to white ip list");
-                result.put("existed", existed);
-                result.put("loaded", loaded);
-                if (!illegalIps.isEmpty()) {
-                    result.put("illegalIps", illegalIps);
+                result.put("existed_ips", existedIPs);
+                result.put("loaded_ips", loadedIPs);
```
Comment (javeme) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpListAPI.java:109]: try `added_xx` or `removed_xx` style

File: hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpListAPI.java [hunk 22]
```diff
@@ -0,0 +1,156 @@
...
+import jakarta.ws.rs.Path;
+import jakarta.ws.rs.Produces;
+import jakarta.ws.rs.QueryParam;
+import jakarta.ws.rs.core.Context;
+
+@Path("whiteiplist")
+@Singleton
+public class WhiteIpListAPI extends API {
+
+    private static final Logger LOG = Log.logger(RestServer.class);
```
Comment (JackyYangPassion) [inline hugegraph-api/src/main/java/org/apache/hugegraph/api/profile/WhiteIpListAPI.java:56]: prefer Log.logger(WhiteIpListAPI.class);

**Review by javeme — DISMISSED** [review]

File: hugegraph-api/src/main/java/org/apache/hugegraph/config/ServerOptions.java [hunk 23]
```diff
@@ -264,4 +264,12 @@ public static synchronized ServerOptions instance() {
...
             );
+
+    public static final ConfigOption<String> WHITE_IP_STATUS =
+            new ConfigOption<>(
+                    "white_ip.status",
+                    "The status of whether enable white ip.",
+                    disallowEmpty(),
+                    "disable"
+            );
 }
```
Comment (imbajin) [inline hugegraph-api/src/main/java/org/apache/hugegraph/config/ServerOptions.java:275]: lack end of line here？

**Review by imbajin — APPROVED** [review]
lgtm

**Review by javeme — APPROVED** [review]

**Pull request merged**